source/meta.tex

```diff
@@ -2432,6 +2434,61 @@
 \end{example}
 \end{itemdescr}
 
+\indexlibraryglobal{is_within_lifetime}%
+\begin{itemdecl}
+template<class T>
+  consteval bool is_within_lifetime(const T* p) noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{true} if \tcode{p} is a pointer to an object that is
+within its lifetime\iref{basic.life}; otherwise, \tcode{false}.
+
+\pnum
+\remarks
+During the evaluation of an expression \tcode{E} as a core constant expression,
+a call to this function is ill-formed
+unless \tcode{p} points to an object that is usable
+in constant expressions or
+whose complete object's lifetime began within \tcode{E}.
+
+\pnum
+\begin{example}
+\begin{codeblock}
+struct OptBool {
+  union { bool b; char c; };
+
+  // note: this assumes common implementation properties for \tcode{bool} and \tcode{char}:
+  // * \tcode{sizeof(bool) == sizeof(char)}, and
+  // * the value representations for \tcode{true} and \tcode{false} are distinct
+  //   from the value representation for \tcode{2}
+  constexpr OptBool() : c(2) { }
+  constexpr OptBool(bool b) : b(b) { }
+
+  constexpr auto has_value() const -> bool {
+    if consteval {
+      return std::is_within_lifetime(&b);       // during constant evaluation, cannot read from \tcode{c}
+    } else {
+      return c != 2;                            // during runtime, must read from \tcode{c}
```

Comment:
This line might have UB. Would it be better to use `std::bit_cast<char>(*this) != 2` instead?

Reply:
You can always read through a char.

Reply:
And the example notes that it relies on common implementation properties. It's a non-normative example, and the committee did not want to clutter it with numerous static assertions to make pedants happy. On common implementations, it works as intended. It's not supposed to be demonstrative of 100% portable code. It's supposed to show how to use `is_within_lifetime`.

Reply:
"Reading through a char" has nothing to do with type punning via a union; it's just reading the object representation, which ought to be valid in all cases. (We have outstanding specification bugs in that area, though.)

Reply:
As I said before, "We have outstanding specification bugs in that area, though.", so "under the current wording" is dubious.  But it's plausible that an lvalue of type "char" refers to the object representation of an object, not necessarily the designated member.

Reply:
Please use the relevant issue trackers for subsequent discussions and clarifications (if any).